### Placement Data Univarient Analysis

#### ✅ Step 1: Import Libraries

```python
import pandas as pd
import numpy as np
from Univariate import Univariate
```

* Imports:

  * `pandas` as `pd` for data manipulation.
  * `numpy` as `np` for numerical operations.
  * `Univariate` class from the `Univariate.py` file you uploaded.

---

In [1]:
import pandas as pd

#### ✅ Step 2: Load the Dataset

```python
df = pd.read_csv("D:/Ram/DS/Pilot_Projects/1.DataUnderstanding/DSTrain.csv")
df.shape
```

* Reads a CSV file into a DataFrame `df`.
* `df.shape` shows the number of rows and columns.

---

In [2]:
dataset=pd.read_csv("Placement.csv")

In [3]:
dataset

sl_no gender  ssc_p    ssc_b  hsc_p    hsc_b     hsc_s  degree_p  \
0        1      M  67.00   Others  91.00   Others  Commerce     58.00   
1        2      M  79.33  Central  78.33   Others   Science     77.48   
2        3      M  65.00  Central  68.00  Central      Arts     64.00   
3        4      M  56.00  Central  52.00  Central   Science     52.00   
4        5      M  85.80  Central  73.60  Central  Commerce     73.30   
..     ...    ...    ...      ...    ...      ...       ...       ...   
210    211      M  80.60   Others  82.00   Others  Commerce     77.60   
211    212      M  58.00   Others  60.00   Others   Science     72.00   
212    213      M  67.00   Others  67.00   Others  Commerce     73.00   
213    214      F  74.00   Others  66.00   Others  Commerce     58.00   
214    215      M  62.00  Central  58.00   Others   Science     53.00   

      degree_t workex  etest_p specialisation  mba_p      status    salary  
0     Sci&Tech     No     55.0         Mkt&HR  58.80      Placed  270000.0  
1     Sci&Tech    Yes     86.5        Mkt&Fin  66.28      Placed  200000.0  
2    Comm&Mgmt     No     75.0        Mkt&Fin  57.80      Placed  250000.0  
3     Sci&Tech     No     66.0         Mkt&HR  59.43  Not Placed       NaN  
4    Comm&Mgmt     No     96.8        Mkt&Fin  55.50      Placed  425000.0  
..         ...    ...      ...            ...    ...         ...       ...  
210  Comm&Mgmt     No     91.0        Mkt&Fin  74.49      Placed  400000.0  
211   Sci&Tech     No     74.0        Mkt&Fin  53.62      Placed  275000.0  
212  Comm&Mgmt    Yes     59.0        Mkt&Fin  69.72      Placed  295000.0  
213  Comm&Mgmt     No     70.0         Mkt&HR  60.23      Placed  204000.0  
214  Comm&Mgmt     No     89.0         Mkt&HR  60.22  Not Placed       NaN  

[215 rows x 15 columns]

In [4]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 215 entries, 0 to 214
Data columns (total 15 columns):
sl_no             215 non-null int64
gender            215 non-null object
ssc_p             215 non-null float64
ssc_b             215 non-null object
hsc_p             215 non-null float64
hsc_b             215 non-null object
hsc_s             215 non-null object
degree_p          215 non-null float64
degree_t          215 non-null object
workex            215 non-null object
etest_p           215 non-null float64
specialisation    215 non-null object
mba_p             215 non-null float64
status            215 non-null object
salary            148 non-null float64
dtypes: float64(6), int64(1), object(8)
memory usage: 25.3+ KB


In [5]:
#Quan=["ssc_p","hsc_p","etest_p","degree_p","mba_p","salary"]
#Qual=["gender","ssc_b","hsc_b","degree_t"]

In [6]:
dataset.columns

Index(['sl_no', 'gender', 'ssc_p', 'ssc_b', 'hsc_p', 'hsc_b', 'hsc_s',
       'degree_p', 'degree_t', 'workex', 'etest_p', 'specialisation', 'mba_p',
       'status', 'salary'],
      dtype='object')

In [7]:
dir(dataset)

['T',
 '_AXIS_ALIASES',
 '_AXIS_IALIASES',
 '_AXIS_LEN',
 '_AXIS_NAMES',
 '_AXIS_NUMBERS',
 '_AXIS_ORDERS',
 '_AXIS_REVERSED',
 '__abs__',
 '__add__',
 '__and__',
 '__array__',
 '__array_priority__',
 '__array_wrap__',
 '__bool__',
 '__class__',
 '__contains__',
 '__copy__',
 '__deepcopy__',
 '__delattr__',
 '__delitem__',
 '__dict__',
 '__dir__',
 '__div__',
 '__doc__',
 '__eq__',
 '__finalize__',
 '__floordiv__',
 '__format__',
 '__ge__',
 '__getattr__',
 '__getattribute__',
 '__getitem__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__iadd__',
 '__iand__',
 '__ifloordiv__',
 '__imod__',
 '__imul__',
 '__init__',
 '__init_subclass__',
 '__invert__',
 '__ior__',
 '__ipow__',
 '__isub__',
 '__iter__',
 '__itruediv__',
 '__ixor__',
 '__le__',
 '__len__',
 '__lt__',
 '__matmul__',
 '__mod__',
 '__module__',
 '__mul__',
 '__ne__',
 '__neg__',
 '__new__',
 '__nonzero__',
 '__or__',
 '__pos__',
 '__pow__',
 '__radd__',
 '__rand__',
 '__rdiv__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 

In [8]:
dataset["ssc_p"].dtypes

dtype('float64')

In [9]:
for columnName in dataset.columns:
    print(columnName)

sl_no
gender
ssc_p
ssc_b
hsc_p
hsc_b
hsc_s
degree_p
degree_t
workex
etest_p
specialisation
mba_p
status
salary


In [10]:
if(dataset["ssc_p"].dtype=='O'):
    print("qual")
else:
    print("quan")

quan


#### ✅ Step 3: Identify Quantitative and Qualitative Columns

* Calls the `quanQual` method to separate columns into:

  * `quan` = list of numeric columns
  * `qual` = list of categorical columns

---

In [11]:
quan=[]
qual=[]
for columnName in dataset.columns:
    #print(columnName)
    if(dataset[columnName].dtype=='O'):
        #print("qual")
        qual.append(columnName)
    else:
        #print("quan")
        quan.append(columnName)
        


In [12]:
quan

['sl_no', 'ssc_p', 'hsc_p', 'degree_p', 'etest_p', 'mba_p', 'salary']

In [13]:
qual

['gender',
 'ssc_b',
 'hsc_b',
 'hsc_s',
 'degree_t',
 'workex',
 'specialisation',
 'status']

#### Explanation:

* **Purpose:**
  The function `quanQual` is designed to separate the columns of a dataset (like a Pandas DataFrame) into two types:

  * **Quantitative (`quan`)**: Numeric data
  * **Qualitative (`qual`)**: Categorical data

* **Step-by-step:**

  1. Takes a DataFrame `dataset` as input.
  2. Initializes two empty lists: `quan` and `qual`.
  3. Iterates over each column name.
  4. Checks the data type of each column:

     * If the type is `'O'` (which means `object` in pandas, commonly used for strings/categories), it's **qualitative**.
     * Otherwise, it's considered **quantitative**.
  5. Appends the column names to the appropriate list.
  6. Returns both lists.

---

In [14]:
def quanQual(dataset):
    quan=[]
    qual=[]
    for columnName in dataset.columns:
        #print(columnName)
        if(dataset[columnName].dtype=='O'):
            #print("qual")
            qual.append(columnName)
        else:
            #print("quan")
            quan.append(columnName)
    return quan,qual


    

In [15]:
quanQual(dataset)

(['sl_no', 'ssc_p', 'hsc_p', 'degree_p', 'etest_p', 'mba_p', 'salary'],
 ['gender',
  'ssc_b',
  'hsc_b',
  'hsc_s',
  'degree_t',
  'workex',
  'specialisation',
  'status'])

In [16]:
quan,qual=quanQual(dataset)

In [17]:
quan

['sl_no', 'ssc_p', 'hsc_p', 'degree_p', 'etest_p', 'mba_p', 'salary']

In [18]:
qual

['gender',
 'ssc_b',
 'hsc_b',
 'hsc_s',
 'degree_t',
 'workex',
 'specialisation',
 'status']

In [19]:
dataset

sl_no gender  ssc_p    ssc_b  hsc_p    hsc_b     hsc_s  degree_p  \
0        1      M  67.00   Others  91.00   Others  Commerce     58.00   
1        2      M  79.33  Central  78.33   Others   Science     77.48   
2        3      M  65.00  Central  68.00  Central      Arts     64.00   
3        4      M  56.00  Central  52.00  Central   Science     52.00   
4        5      M  85.80  Central  73.60  Central  Commerce     73.30   
..     ...    ...    ...      ...    ...      ...       ...       ...   
210    211      M  80.60   Others  82.00   Others  Commerce     77.60   
211    212      M  58.00   Others  60.00   Others   Science     72.00   
212    213      M  67.00   Others  67.00   Others  Commerce     73.00   
213    214      F  74.00   Others  66.00   Others  Commerce     58.00   
214    215      M  62.00  Central  58.00   Others   Science     53.00   

      degree_t workex  etest_p specialisation  mba_p      status    salary  
0     Sci&Tech     No     55.0         Mkt&HR  58.80      Placed  270000.0  
1     Sci&Tech    Yes     86.5        Mkt&Fin  66.28      Placed  200000.0  
2    Comm&Mgmt     No     75.0        Mkt&Fin  57.80      Placed  250000.0  
3     Sci&Tech     No     66.0         Mkt&HR  59.43  Not Placed       NaN  
4    Comm&Mgmt     No     96.8        Mkt&Fin  55.50      Placed  425000.0  
..         ...    ...      ...            ...    ...         ...       ...  
210  Comm&Mgmt     No     91.0        Mkt&Fin  74.49      Placed  400000.0  
211   Sci&Tech     No     74.0        Mkt&Fin  53.62      Placed  275000.0  
212  Comm&Mgmt    Yes     59.0        Mkt&Fin  69.72      Placed  295000.0  
213  Comm&Mgmt     No     70.0         Mkt&HR  60.23      Placed  204000.0  
214  Comm&Mgmt     No     89.0         Mkt&HR  60.22  Not Placed       NaN  

[215 rows x 15 columns]

# Central tendency

**Markdown note for Central Tendency**, written in the same structured style as the Percentile Analysis, but **focused entirely on your placement dataset** and using your provided mean, median, and mode values.

---

### 📘 **Central Tendency – Placement Data Analysis**

Central tendency is a statistical concept used to identify the **center point or typical value** in a dataset. It helps summarize the overall performance or behavior of a group using three key metrics: **mean**, **median**, and **mode**.

---

### 🔑 **Key Metrics**

| Metric     | Description                         |
| ---------- | ----------------------------------- |
| **Mean**   | The average of all values           |
| **Median** | The middle value when sorted        |
| **Mode**   | The most frequently occurring value |

---

### 📊 **Central Tendency for Student Placement Data**

| Metric     | SSC (%) | HSC (%) | Degree (%) | E-test (%) | MBA (%) | Salary (INR) |
| ---------- | ------- | ------- | ---------- | ---------- | ------- | ------------ |
| **Mean**   | 67.30   | 66.33   | 66.37      | 72.10      | 62.28   | 288,655      |
| **Median** | 67.00   | 65.00   | 66.00      | 71.00      | 62.00   | 265,000      |
| **Mode**   | 62.00   | 63.00   | 65.00      | 60.00      | 56.70   | 300,000      |

---

### 🧠 **Interpretation**

* All **mean and median values** for academic performance fall between **60% and 79%**, indicating that **most students performed at an average level** based on your grading criteria.
* **E-test** scores have the highest average (72.10%), which may reflect aptitude or general test-taking ability.
* The **salary mean (₹288,655)** is higher than the **median (₹265,000)**, suggesting a **right-skewed distribution**, where a few students may have received very high packages.

---

### 📐 **Individual Metric Meaning**

* A student with a **degree % = 65%** is very close to the **median**, which means their performance is typical of the group.
* A salary of ₹300,000 corresponds to the **mode**, suggesting that multiple students received this amount.

---

### ✅ **Final Verdict: Average or Good?**

Based on your criteria:

* **60%–79% = Average**
* **80%+ = Good**

Since most central tendency values fall in the **60–79% range**, we conclude:

> 🟢 **"The students in this dataset are, on average, performing at an average level across academic and aptitude indicators, but still achieving decent salary outcomes in placements."**



In [20]:
dataset["ssc_p"].mean()

67.30339534883721

In [21]:
dataset["ssc_p"].median()

67.0

In [22]:
dataset["ssc_p"].mode()[0]

62.0

In [23]:
dataset.describe()

sl_no       ssc_p       hsc_p    degree_p     etest_p       mba_p  \
count  215.000000  215.000000  215.000000  215.000000  215.000000  215.000000   
mean   108.000000   67.303395   66.333163   66.370186   72.100558   62.278186   
std     62.209324   10.827205   10.897509    7.358743   13.275956    5.833385   
min      1.000000   40.890000   37.000000   50.000000   50.000000   51.210000   
25%     54.500000   60.600000   60.900000   61.000000   60.000000   57.945000   
50%    108.000000   67.000000   65.000000   66.000000   71.000000   62.000000   
75%    161.500000   75.700000   73.000000   72.000000   83.500000   66.255000   
max    215.000000   89.400000   97.700000   91.000000   98.000000   77.890000   

              salary  
count     148.000000  
mean   288655.405405  
std     93457.452420  
min    200000.000000  
25%    240000.000000  
50%    265000.000000  
75%    300000.000000  
max    940000.000000

In [24]:
descriptive=pd.DataFrame(index=["Mean","Median","Mode"],columns=quan)

In [25]:
descriptive["sl_no"]["Mean"]

nan

In [26]:
descriptive=pd.DataFrame(index=["Mean","Median","Mode"],columns=quan)
for columnName in quan:
    descriptive[columnName]["Mean"]=dataset[columnName].mean()

In [27]:
descriptive

sl_no    ssc_p    hsc_p degree_p  etest_p    mba_p  salary
Mean     108  67.3034  66.3332  66.3702  72.1006  62.2782  288655
Median   NaN      NaN      NaN      NaN      NaN      NaN     NaN
Mode     NaN      NaN      NaN      NaN      NaN      NaN     NaN

In [28]:
descriptive=pd.DataFrame(index=["Mean","Median","Mode"],columns=quan)
for columnName in quan:
    descriptive[columnName]["Mean"]=dataset[columnName].mean()
    descriptive[columnName]["Median"]=dataset[columnName].median()
    descriptive[columnName]["Mode"]=dataset[columnName].mode()[0]

In [29]:
descriptive

sl_no    ssc_p    hsc_p degree_p  etest_p    mba_p  salary
Mean     108  67.3034  66.3332  66.3702  72.1006  62.2782  288655
Median   108       67       65       66       71       62  265000
Mode       1       62       63       65       60     56.7  300000


### 📊 **Placement Data Analysis – Student Performance and Salary Insights**

The dataset represents the academic and placement performance of **108 students**, focusing on their percentage scores and salaries after placement. Here’s a summary based on the **central tendency (mean, median, and mode)** values of each column:

#### 🧠 **Academic Performance (Percentages)**

| Metric     | SSC (%) | HSC (%) | Degree (%) | E-Test (%) | MBA (%) |
| ---------- | ------- | ------- | ---------- | ---------- | ------- |
| **Mean**   | 67.30   | 66.33   | 66.37      | 72.10      | 62.28   |
| **Median** | 67.00   | 65.00   | 66.00      | 71.00      | 62.00   |
| **Mode**   | 62.00   | 63.00   | 65.00      | 60.00      | 56.70   |

* All **mean and median values** lie between **60% and 79%**, indicating an **average academic performance** across SSC, HSC, Degree, E-test, and MBA.
* **E-test performance (mean = 72.10%)** stands out as relatively strong among the students.
* The **mode** values reflect the most frequently occurring scores, which are slightly lower, suggesting some students scored on the lower average side in repeated patterns.

#### 💰 **Salary Insights**

| Metric     | Salary (INR) |
| ---------- | ------------ |
| **Mean**   | 288,655      |
| **Median** | 265,000      |
| **Mode**   | 300,000      |

* The **average salary** offered to students is around **₹288,655**, with the **median at ₹265,000**, indicating most students received offers in that range.
* A **mode of ₹300,000** suggests a common offer amount was around 3 Lakhs, which could represent a standard package from one or more recruiters.

---

### ✅ **Final Conclusion**

Based on the provided data and your criteria:

* Students scoring **above 60% and below 80%** are considered **average**.
* Since **all academic averages** (mean, median) lie within this range, we conclude that:

> 🟢 **"The overall academic performance of students is average. However, the salary figures suggest a healthy placement trend, indicating employability despite average academic scores."**



In [30]:
import numpy as np

In [31]:
np.percentile(dataset["ssc_p"],50)

67.0

**clean and informative Markdown note** you can use in your Jupyter notebook to understand the concept of **percentiles**, along with how to apply it in data analysis.

---

### 📘 **Understanding Percentiles in Data Analysis**

#### 🔢 What is a Percentile?

A **percentile** is a measure used in statistics to indicate the value below which a given percentage of observations in a group falls or value exists within the range.

It dividing in parts of intial vale,25%(Q1),50%(Q2),75%(Q3),100%(Q1)Final Value.

#### 🔢 What is a Purpose of Percentile?
* Sort the data.
* Asending the order of the data.

#### 🔢Formula: Percentile (i)=k/100*(n+1) (Position of data)

For example:

* The **25th percentile** is the value below which 25% of the data lies.
* The **50th percentile** is the **median**.
* The **90th percentile** is the value below which 90% of the data lies — only the top 10% exceed it.

---

#### 🧠 **Why Are Percentiles Useful?**

Percentiles help in:

* Understanding **data distribution**
* Identifying **outliers**
* Comparing **individual values** against the whole dataset (e.g., "This student scored higher than 80% of their peers")

---

#### 🧮 **How to Calculate Percentiles in Python**

```python
import numpy as np

# Example: Calculate the 25th, 50th, and 75th percentiles of a column
percentiles = np.percentile(df['ssc_p'], [25, 50, 75])
print("25th:", percentiles[0], "50th (Median):", percentiles[1], "75th:", percentiles[2])
```

---

#### 📏 **Percentile vs Percentile Rank**

* **Percentile Value**: Given a percentile (e.g., 90th), find the score at that point.
* **Percentile Rank**: Given a score, find what percentile it belongs to.

---

#### 📈 **Use Case in Student Data Analysis**

By calculating percentiles for academic percentages and salaries, we can:

* Identify **average vs top performers**
* Detect **inequality or skew** in placement offers
* Profile students (e.g., top 10% scorers)

In [32]:
dataset.describe()

sl_no       ssc_p       hsc_p    degree_p     etest_p       mba_p  \
count  215.000000  215.000000  215.000000  215.000000  215.000000  215.000000   
mean   108.000000   67.303395   66.333163   66.370186   72.100558   62.278186   
std     62.209324   10.827205   10.897509    7.358743   13.275956    5.833385   
min      1.000000   40.890000   37.000000   50.000000   50.000000   51.210000   
25%     54.500000   60.600000   60.900000   61.000000   60.000000   57.945000   
50%    108.000000   67.000000   65.000000   66.000000   71.000000   62.000000   
75%    161.500000   75.700000   73.000000   72.000000   83.500000   66.255000   
max    215.000000   89.400000   97.700000   91.000000   98.000000   77.890000   

              salary  
count     148.000000  
mean   288655.405405  
std     93457.452420  
min    200000.000000  
25%    240000.000000  
50%    265000.000000  
75%    300000.000000  
max    940000.000000

In [33]:
descriptive=pd.DataFrame(index=["Mean","Median","Mode","25%","50%","75%","99%","100%"],columns=quan)
for columnName in quan:
    descriptive[columnName]["Mean"]=dataset[columnName].mean()
    descriptive[columnName]["Median"]=dataset[columnName].median()
    descriptive[columnName]["Mode"]=dataset[columnName].mode()[0]
    descriptive[columnName]["25%"]=dataset.describe()[columnName]["25%"]
    descriptive[columnName]["50%"]=dataset.describe()[columnName]["50%"]
    descriptive[columnName]["75%"]=dataset.describe()[columnName]["75%"]
    descriptive[columnName]["99%"]=np.percentile(dataset[columnName],99)
    descriptive[columnName]["100%"]=dataset.describe()[columnName]["max"]

In [34]:
descriptive

sl_no    ssc_p    hsc_p degree_p  etest_p    mba_p  salary
Mean       108  67.3034  66.3332  66.3702  72.1006  62.2782  288655
Median     108       67       65       66       71       62  265000
Mode         1       62       63       65       60     56.7  300000
25%       54.5     60.6     60.9       61       60   57.945  240000
50%        108       67       65       66       71       62  265000
75%      161.5     75.7       73       72     83.5   66.255  300000
99%     212.86       87    91.86    83.86       97  76.1142     NaN
100%       215     89.4     97.7       91       98    77.89  940000


## 📊 Data Analysis Report by Departmental Columns

This report provides a detailed summary for each quantitative variable based on **central tendency (Mean, Median, Mode)** and **percentile ranges (25%, 50%, 75%, 99%, 100%)**. The goal is to help your client understand **how data is distributed**, **where the data concentrates**, and **where extreme or exceptional values exist**.

---

### 1. **Secondary School Percentage (`ssc_p`)**

* **Mean:** 67.30% – the average score across all candidates.
* **Median:** 67% – half of the candidates scored below this.
* **Mode:** 62% – most frequently occurring score.

#### 📈 Percentile Range Differences:

* **25% to 50%:** 67 - 60.6 = **6.4%** → Most students scored moderately in this range.
* **50% to 75%:** 75.7 - 67 = **8.7%** → Slight increase in performance among higher scorers.
* **75% to 99%:** 87 - 75.7 = **11.3%** → High-performing candidates, upward trend.
* **99% to 100%:** 89.4 - 87 = **2.4%** → Minimal jump, indicating a ceiling effect (max scores don’t vary much beyond top 1%).

---

### 2. **High School Percentage (`hsc_p`)**

* **Mean:** 66.33%
* **Median:** 65%
* **Mode:** 63%

#### 📈 Percentile Range Differences:

* **25% to 50%:** 65 - 60.9 = **4.1%** → Lower-middle scoring concentration.
* **50% to 75%:** 73 - 65 = **8%** → Strong improvement in upper half.
* **75% to 99%:** 91.86 - 73 = **18.86%** → High variance among top scorers.
* **99% to 100%:** 97.7 - 91.86 = **5.84%** → Exceptional performers with very high scores.

---

### 3. **Degree Percentage (`degree_p`)**

* **Mean:** 66.37%
* **Median:** 66%
* **Mode:** 65%

#### 📈 Percentile Range Differences:

* **25% to 50%:** 66 - 61 = **5%**
* **50% to 75%:** 72 - 66 = **6%**
* **75% to 99%:** 83.86 - 72 = **11.86%**
* **99% to 100%:** 91 - 83.86 = **7.14%**

**Insight:** Degree scores have a **well-balanced distribution** with modest central values and some notable jump in top percentiles, showing a few standout students.

---

### 4. **Employability Test Percentage (`etest_p`)**

* **Mean:** 72.10%
* **Median:** 71%
* **Mode:** 60%

#### 📈 Percentile Range Differences:

* **25% to 50%:** 71 - 60 = **11%**
* **50% to 75%:** 83.5 - 71 = **12.5%**
* **75% to 99%:** 97 - 83.5 = **13.5%**
* **99% to 100%:** 98 - 97 = **1%**

**Insight:** This metric has the **largest growth across percentile bands**, especially in the middle and upper percentiles, indicating that top candidates show **markedly better employability test scores**.

---

### 5. **MBA Percentage (`mba_p`)**

* **Mean:** 62.28%
* **Median:** 62%
* **Mode:** 56.7%

#### 📈 Percentile Range Differences:

* **25% to 50%:** 62 - 57.945 = **4.055%**
* **50% to 75%:** 66.255 - 62 = **4.255%**
* **75% to 99%:** 76.11 - 66.255 = **9.855%**
* **99% to 100%:** 77.89 - 76.1142 = **1.7758%**

**Insight:** The MBA scores are **tightly packed** below the 75th percentile and only expand noticeably between the 75th and 99th. There's little variation among the very top scorers, suggesting a consistent grading standard.

---

### 6. **Salary (Annual in ₹)**

* **Mean:** ₹288,655
* **Median:** ₹265,000
* **Mode:** ₹300,000

#### 📈 Percentile Range Differences:

* **25% to 50%:** ₹265,000 - ₹240,000 = **₹25,000**
* **50% to 75%:** ₹300,000 - ₹265,000 = **₹35,000**
* **75% to 99%:** *Missing (NaN)* → Data unavailable
* **99% to 100%:** ₹940,000 - *(missing)*

**Insight:** While we can see a moderate increase from 25% to 75%, the **lack of a 99th percentile value** suggests missing data in that band. However, the **maximum salary is significantly higher (₹940,000)**, which clearly indicates the presence of **strong outliers or elite placement packages**.

---

## 📌😊🧠 Strategic Insights for the Client:

1. **ssc\_p, hsc\_p, and degree\_p** show steady and expected growth between percentile bands — with a few top performers significantly ahead.
2. **etest\_p** is the **most telling feature** in differentiating high performers, especially in the upper percentile ranges.
3. **mba\_p** has limited spread, which might suggest consistent teaching/grading or limited differentiation.
4. **salary** data confirms **skewed distribution with extreme outliers**, typical in job markets where a few land top-tier packages.

---

In [35]:
descriptive=pd.DataFrame(index=["Mean","Median","Mode","Min","25%","50%","75%","100%"],columns=quan)
for columnName in quan:
    descriptive[columnName]["Mean"]=dataset[columnName].mean()
    descriptive[columnName]["Median"]=dataset[columnName].median()
    descriptive[columnName]["Mode"]=dataset[columnName].mode()[0]
    descriptive[columnName]["Min"]=dataset[columnName].min()
    descriptive[columnName]["25%"]=dataset.describe()[columnName]["25%"]
    descriptive[columnName]["50%"]=dataset.describe()[columnName]["50%"]
    descriptive[columnName]["75%"]=dataset.describe()[columnName]["75%"]
    descriptive[columnName]["75%"]=dataset.describe()[columnName]["75%"]
    descriptive[columnName]["100%"]=dataset.describe()[columnName]["max"]

In [36]:
descriptive

sl_no    ssc_p    hsc_p degree_p  etest_p    mba_p  salary
Mean      108  67.3034  66.3332  66.3702  72.1006  62.2782  288655
Median    108       67       65       66       71       62  265000
Mode        1       62       63       65       60     56.7  300000
Min         1    40.89       37       50       50    51.21  200000
25%      54.5     60.6     60.9       61       60   57.945  240000
50%       108       67       65       66       71       62  265000
75%     161.5     75.7       73       72     83.5   66.255  300000
100%      215     89.4     97.7       91       98    77.89  940000

In [37]:
descriptive=pd.DataFrame(index=["Mean","Median","Mode","Min","Max","25%","50%","75%","100%","IQR",
                               "1.5IQR","LesserRange","GreaterRange"],columns=quan)
for columnName in quan:
    descriptive[columnName]["Mean"]=dataset[columnName].mean()
    descriptive[columnName]["Median"]=dataset[columnName].median()
    descriptive[columnName]["Mode"]=dataset[columnName].mode()[0]
    descriptive[columnName]["Min"]=dataset[columnName].min()
    descriptive[columnName]["Max"]=dataset[columnName].max()
    descriptive[columnName]["25%"]=dataset.describe()[columnName]["25%"]
    descriptive[columnName]["50%"]=dataset.describe()[columnName]["50%"]
    descriptive[columnName]["75%"]=dataset.describe()[columnName]["75%"]
    descriptive[columnName]["100%"]=dataset.describe()[columnName]["max"]
    descriptive[columnName]["IQR"]=descriptive[columnName]["75%"]-descriptive[columnName]["25%"]
    descriptive[columnName]["1.5IQR"]=1.5*descriptive[columnName]["IQR"]
    descriptive[columnName]["LesserRange"]=descriptive[columnName]["25%"]- descriptive[columnName]["1.5IQR"]
    descriptive[columnName]["GreaterRange"]=descriptive[columnName]["75%"]+ descriptive[columnName]["1.5IQR"]


In [38]:
descriptive

sl_no    ssc_p    hsc_p degree_p  etest_p    mba_p  salary
Mean            108  67.3034  66.3332  66.3702  72.1006  62.2782  288655
Median          108       67       65       66       71       62  265000
Mode              1       62       63       65       60     56.7  300000
Min               1    40.89       37       50       50    51.21  200000
Max             215     89.4     97.7       91       98    77.89  940000
25%            54.5     60.6     60.9       61       60   57.945  240000
50%             108       67       65       66       71       62  265000
75%           161.5     75.7       73       72     83.5   66.255  300000
100%            215     89.4     97.7       91       98    77.89  940000
IQR             107     15.1     12.1       11     23.5     8.31   60000
1.5IQR        160.5    22.65    18.15     16.5    35.25   12.465   90000
LesserRange    -106    37.95    42.75     44.5    24.75    45.48  150000
GreaterRange    322    98.35    91.15     88.5   118.75    78.72  390000

# Data Analysis Report: IQR Parameters

## 1. Interquartile Range (IQR)
- **Definition**: IQR measures the spread of the middle 50% of the data (Q3 - Q1).
- **Values**:
  - `ssc_p`: 15.1  
  - `hsc_p`: 12.1  
  - `degree_p`: 11  
  - `etest_p`: 23.5  
  - `mba_p`: 8.31  
  - `salary`: 60,000  

## 2. 1.5 × IQR Rule (Outlier Detection)
- **Definition**: Values below `Q1 - 1.5IQR` or above `Q3 + 1.5IQR` are potential outliers.
- **Calculated Ranges**:
  - **Lower Bound (Q1 - 1.5IQR)**:
    - `ssc_p`: 37.95  
    - `hsc_p`: 42.75  
    - `degree_p`: 44.5  
    - `etest_p`: 24.75  
    - `mba_p`: 45.48  
    - `salary`: 150,000  
  - **Upper Bound (Q3 + 1.5IQR)**:
    - `ssc_p`: 98.35  
    - `hsc_p`: 91.15  
    - `degree_p`: 88.5  
    - `etest_p`: 118.75  
    - `mba_p`: 78.72  
    - `salary`: 390,000  

## 3. Min and Max Values
- **Min**: The smallest observed value in the dataset.
  - Example: `salary` starts at 200,000 (no values below this).
- **Max**: The largest observed value in the dataset.
  - Example: `salary` peaks at 940,000 (exceeds upper bound → **potential outlier**).

## 4. Lesser and Greater Ranges
- **Lesser Range**: Theoretical minimum after outlier adjustment (`Q1 - 1.5IQR`).
  - Example: `etest_p` suggests values <24.75 are outliers, but actual `Min=50` → **no outliers** here.
- **Greater Range**: Theoretical maximum after outlier adjustment (`Q3 + 1.5IQR`).
  - Example: `salary` has `Max=940,000` > `GreaterRange=390,000` → **high outliers exist**.

## Key Insights
- **Outliers**: `salary` has significant outliers on the higher end (e.g., 940,000 vs. 390,000 upper bound).
- **Tight Spread**: `mba_p` has a small IQR (8.31), indicating consistent performance.
- **Wide Spread**: `etest_p` has the largest IQR (23.5), suggesting varied scores.

In [39]:

if(descriptive["ssc_p"]["LesserRange"]>descriptive["ssc_p"]["Min"]):
    print("LesserRange Outlier")
else:
    print("No LR Outlier")

No LR Outlier


In [40]:
lesser=[]
greater=[]

for columnName in quan:
    if(descriptive[columnName]["LesserRange"]>descriptive[columnName]["Min"]):
        lesser.append(columnName)
    if(descriptive[columnName]["GreaterRange"]<descriptive[columnName]["100%"]):
        greater.append(columnName)

In [41]:
lesser

['hsc_p']

In [42]:
greater

['hsc_p', 'degree_p', 'salary']

In [43]:
def Outlier_Column_Names(descriptive,quan):
    lesser=[]
    greater=[]
    for columnName in quan:
        if(descriptive[columnName]["LesserRange"]>descriptive[columnName]["Min"]):
            lesser.append(columnName)
        if(descriptive[columnName]["GreaterRange"]<descriptive[columnName]["100%"]):
            greater.append(columnName)
    return lesser,greater

In [44]:
Outlier_Column_Names(descriptive,quan)

(['hsc_p'], ['hsc_p', 'degree_p', 'salary'])

In [45]:
#Once we find out the outlier next step change the range of outlier in dataset.
for lesscolumn in lesser:
    dataset[lesscolumn][dataset[lesscolumn]<descriptive[lesscolumn]["LesserRange"]]=descriptive[lesscolumn]["LesserRange"]
for greatcolumn in greater:
    dataset[greatcolumn][dataset[greatcolumn]>descriptive[greatcolumn]["GreaterRange"]]=descriptive[greatcolumn]["GreaterRange"]


C:\Anaconda3\envs\aiml\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Anaconda3\envs\aiml\lib\site-packages\ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """


In [46]:
def Replace_Outlier(descriptive, dataset):
    
    for lesscolumn in lesser:
        dataset[lesscolumn][dataset[lesscolumn]<descriptive[lesscolumn]["LesserRange"]]=descriptive[lesscolumn]["LesserRange"]
    for greatcolumn in greater:
        dataset[greatcolumn][dataset[greatcolumn]>descriptive[greatcolumn]["GreaterRange"]]=descriptive[greatcolumn]["GreaterRange"]
    return columnName

Replace_Outlier(descriptive, dataset)

C:\Anaconda3\envs\aiml\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.
C:\Anaconda3\envs\aiml\lib\site-packages\ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


'salary'

In [47]:
def Univariate(dataset,quan):
    descriptive=pd.DataFrame(index=["Mean","Median","Mode","Min","Max","25%","50%","75%","100%","IQR",
                               "1.5IQR","LesserRange","GreaterRange","Skewness","Kurtosis"],columns=quan)
    
    for columnName in quan:
        descriptive[columnName]["Mean"]=dataset[columnName].mean()
        descriptive[columnName]["Median"]=dataset[columnName].median()
        descriptive[columnName]["Mode"]=dataset[columnName].mode()[0]
        descriptive[columnName]["Min"]=dataset[columnName].min()
        descriptive[columnName]["Max"]=dataset[columnName].max()
        descriptive[columnName]["25%"]=dataset.describe()[columnName]["25%"]
        descriptive[columnName]["50%"]=dataset.describe()[columnName]["50%"]
        descriptive[columnName]["75%"]=dataset.describe()[columnName]["75%"]
        descriptive[columnName]["100%"]=dataset.describe()[columnName]["max"]
        descriptive[columnName]["IQR"]=descriptive[columnName]["75%"]-descriptive[columnName]["25%"]
        descriptive[columnName]["1.5IQR"]=1.5*descriptive[columnName]["IQR"]
        descriptive[columnName]["LesserRange"]=descriptive[columnName]["25%"]- descriptive[columnName]["1.5IQR"]
        descriptive[columnName]["GreaterRange"]=descriptive[columnName]["75%"]+ descriptive[columnName]["1.5IQR"]
        descriptive[columnName]["Skewness"]=dataset[columnName].skew()
        descriptive[columnName]["Kurtosis"]=dataset[columnName].kurtosis()
    return descriptive

Univariate(dataset,quan)

sl_no     ssc_p      hsc_p   degree_p   etest_p     mba_p  \
Mean            108   67.3034    66.3347    66.3586   72.1006   62.2782   
Median          108        67         65         66        71        62   
Mode              1        62         63         65        60      56.7   
Min               1     40.89      42.75         50        50     51.21   
Max             215      89.4      91.15       88.5        98     77.89   
25%            54.5      60.6       60.9         61        60    57.945   
50%             108        67         65         66        71        62   
75%           161.5      75.7         73         72      83.5    66.255   
100%            215      89.4      91.15       88.5        98     77.89   
IQR             107      15.1       12.1         11      23.5      8.31   
1.5IQR        160.5     22.65      18.15       16.5     35.25    12.465   
LesserRange    -106     37.95      42.75       44.5     24.75     45.48   
GreaterRange    322     98.35      91.15       88.5    118.75     78.72   
Skewness          0 -0.132649   0.162611   0.204164  0.282308  0.313576   
Kurtosis       -1.2  -0.60751  0.0869008 -0.0974897  -1.08858 -0.470723   

                salary  
Mean            277649  
Median          265000  
Mode            300000  
Min             200000  
Max             390000  
25%             240000  
50%             265000  
75%             300000  
100%            390000  
IQR              60000  
1.5IQR           90000  
LesserRange     150000  
GreaterRange    390000  
Skewness        0.8067  
Kurtosis     -0.239837

# Descriptive Statistics Calculation Code Breakdown

## **1. DataFrame Initialization**
```python
descriptive = pd.DataFrame(
    index=["Mean", "Median", "Mode", "Min", "25%", "50%", "75%", "100%", "IQR", "1.5IQR", "LesserRange", "GreaterRange"],
    columns=quan
)
```
- **Purpose**: Creates an empty DataFrame with:  
  - **Rows**: Statistical measures (e.g., mean, median, IQR).  
  - **Columns**: Numerical variables (from `quan`).  

---

## **2. Loop Through Numerical Columns**
```python
for columnName in quan:
```
- Iterates over each numerical column (e.g., `salary`, `mba_p`) to compute statistics.

---

## **3. Core Statistics Calculation**
### **Central Tendency**
| Code | Statistic | Description |
|------|-----------|-------------|
| `dataset[columnName].mean()` | Mean | Average value |
| `dataset[columnName].median()` | Median | Middle value (50th percentile) |
| `dataset[columnName].mode()[0]` | Mode | Most frequent value |

### **Spread & Extremes**
| Code | Statistic | Description |
|------|-----------|-------------|
| `dataset[columnName].min()` | Min | Smallest value |
| `describe()[columnName]["25%"]` | 25th percentile (Q1) | Lower quartile |
| `describe()[columnName]["75%"]` | 75th percentile (Q3) | Upper quartile |
| `describe()[columnName]["max"]` | Max | Largest value |

---

## **4. Interquartile Range (IQR)**
```python
descriptive[columnName]["IQR"] = Q3 - Q1
```
- **IQR** = `75th percentile - 25th percentile`  
- Measures spread of the middle 50% of data.

---

## **5. Outlier Detection (1.5×IQR Rule)**
```python
descriptive[columnName]["1.5IQR"] = 1.5 * IQR
descriptive[columnName]["LesserRange"] = Q1 - 1.5IQR  # Lower outlier threshold
descriptive[columnName]["GreaterRange"] = Q3 + 1.5IQR # Upper outlier threshold
```
- **Outliers**: Values outside `[Q1 - 1.5IQR, Q3 + 1.5IQR]`.

---

## **Key Insights from Output**
- **Wide IQR** (e.g., `etest_p` = 23.5) → High variability in data.  
- **Outliers** (e.g., `salary` max = 940K > `GreaterRange` = 390K) → Extreme high values exist.  
- **Tight Spread** (e.g., `mba_p` IQR = 8.31) → Consistent performance.

---

## **Why This Matters**
- **Data Quality**: Identifies outliers needing review.  
- **Comparative Analysis**: Highlights variables with unusual distributions.  
- **Modeling Prep**: Helps decide normalization/log-transforms.
```

In [48]:
lesser=[]
greater=[]

for columnName in quan:
    if(descriptive[columnName]["LesserRange"]>descriptive[columnName]["Min"]):
        lesser.append(columnName)
    if(descriptive[columnName]["GreaterRange"]<descriptive[columnName]["100%"]):
        greater.append(columnName)

In [49]:
lesser

['hsc_p']

In [50]:
greater

['hsc_p', 'degree_p', 'salary']

In [51]:
descriptive

sl_no    ssc_p    hsc_p degree_p  etest_p    mba_p  salary
Mean            108  67.3034  66.3332  66.3702  72.1006  62.2782  288655
Median          108       67       65       66       71       62  265000
Mode              1       62       63       65       60     56.7  300000
Min               1    40.89       37       50       50    51.21  200000
Max             215     89.4     97.7       91       98    77.89  940000
25%            54.5     60.6     60.9       61       60   57.945  240000
50%             108       67       65       66       71       62  265000
75%           161.5     75.7       73       72     83.5   66.255  300000
100%            215     89.4     97.7       91       98    77.89  940000
IQR             107     15.1     12.1       11     23.5     8.31   60000
1.5IQR        160.5    22.65    18.15     16.5    35.25   12.465   90000
LesserRange    -106    37.95    42.75     44.5    24.75    45.48  150000
GreaterRange    322    98.35    91.15     88.5   118.75    78.72  390000

# Frequency 

## What is a Frequency Table?
A frequency table shows how often different values appear in your data. It helps you understand patterns and distributions.

### Key Components:
1. **Unique Values** - Each different item in your data
2. **Frequency** - How many times each value appears
3. **Relative Frequency** - Percentage of total for each value
4. **Cumulative Sum** - Running total of percentages


## Understanding Frequency Concepts

### 1. Types of Frequency Measures

| Measure | Description | Formula | When to Use |
|---------|-------------|---------|-------------|
| **Frequency** | Raw count of each unique value | Count occurrences | Understanding absolute distribution |
| **Relative Frequency** | Proportion of each value in dataset | Frequency/Total observations | Comparing distributions with different sample sizes |
| **Cumulative Frequency** | Running total of frequencies | Sum of all previous frequencies | Analyzing data distribution patterns |
| **Cumulative Relative Frequency** | Running total of relative frequencies | Sum of all previous relative frequencies | Determining percentiles and quartiles |

### 2. Types of Variables

| Variable Type | Frequency Table Use | Example |
|--------------|---------------------|---------|
| **Categorical** | Most common use case | Gender, Product Categories |
| **Ordinal** | Shows ordered relationships | Education Levels, Satisfaction Ratings |
| **Discrete Numerical** | Works well with limited unique values | Number of children, Shoe sizes |
| **Continuous Numerical** | Often requires binning first | Age groups, Income ranges |

In [52]:
dataset["ssc_p"].value_counts().index

Float64Index([ 62.0,  63.0,  67.0,  52.0,  73.0,  65.0,  74.0,  69.0,  58.0,
               77.0,
              ...
               64.6,  60.8,  63.3,  74.2,  61.8,  80.6,  65.2,  80.4, 59.96,
               78.5],
             dtype='float64', length=103)

In [53]:
dataset["ssc_p"].value_counts().sort_values

<bound method Series.sort_values of 62.00    11
63.00    10
67.00     9
52.00     9
73.00     9
         ..
80.60     1
65.20     1
80.40     1
59.96     1
78.50     1
Name: ssc_p, Length: 103, dtype: int64>

In [54]:
freqTable=pd.DataFrame(columns=["Unique_Values","Frequency","Relative_Frequency","Cumsum"])

In [55]:
freqTable["Unique_Values"]=dataset["ssc_p"].value_counts().index

In [56]:
freqTable

Unique_Values Frequency Relative_Frequency Cumsum
0            62.00       NaN                NaN    NaN
1            63.00       NaN                NaN    NaN
2            67.00       NaN                NaN    NaN
3            52.00       NaN                NaN    NaN
4            73.00       NaN                NaN    NaN
..             ...       ...                ...    ...
98           80.60       NaN                NaN    NaN
99           65.20       NaN                NaN    NaN
100          80.40       NaN                NaN    NaN
101          59.96       NaN                NaN    NaN
102          78.50       NaN                NaN    NaN

[103 rows x 4 columns]

In [57]:
freqTable["Unique_Values"]=dataset["ssc_p"].value_counts().index
freqTable["Frequency"]=dataset["ssc_p"].value_counts().values
freqTable["Relative_Frequency"]=(freqTable["Frequency"]/103)
freqTable["Cumsum"]=freqTable["Relative_Frequency"].cumsum() 

In [58]:
freqTable

Unique_Values  Frequency  Relative_Frequency    Cumsum
0            62.00         11            0.106796  0.106796
1            63.00         10            0.097087  0.203883
2            67.00          9            0.087379  0.291262
3            52.00          9            0.087379  0.378641
4            73.00          9            0.087379  0.466019
..             ...        ...                 ...       ...
98           80.60          1            0.009709  2.048544
99           65.20          1            0.009709  2.058252
100          80.40          1            0.009709  2.067961
101          59.96          1            0.009709  2.077670
102          78.50          1            0.009709  2.087379

[103 rows x 4 columns]

In [59]:
def freqTable(colunName,dataset):
    freqTable=pd.DataFrame(columns=["Unique_Values","Frequency","Relative_Frequency","Cumsum"])
    freqTable["Unique_Values"]=dataset[columnName].value_counts().index
    freqTable["Frequency"]=dataset[columnName].value_counts().values
    freqTable["Relative_Frequency"]=(freqTable["Frequency"]/103)
    freqTable["Cumsum"]=freqTable["Relative_Frequency"].cumsum()
    return freqTable

##  Python Code Explained

```python
def freqTable(columnName, dataset):
    # Create empty table with these columns
    freqTable = pd.DataFrame(columns=[
        "Unique_Values",    # Stores each different value
        "Frequency",        # Counts how often each appears
        "Relative_Frequency", # Shows percentage (count/total)
        "Cumsum"           # Running total of percentages
    ])
    
    # Get all unique values and their counts
    freqTable["Unique_Values"] = dataset[columnName].value_counts().index
    freqTable["Frequency"] = dataset[columnName].value_counts().values
    
    # Calculate percentages (uses 103 as total - should change this!)
    freqTable["Relative_Frequency"] = (freqTable["Frequency"]/103)
    
    # Calculate running percentage total
    freqTable["Cumsum"] = freqTable["Relative_Frequency"].cumsum()
    
    return freqTable

In [60]:
freqTable("hsc_p",dataset)

Unique_Values  Frequency  Relative_Frequency    Cumsum
0        300000.0         22            0.213592  0.213592
1        250000.0         18            0.174757  0.388350
2        390000.0         15            0.145631  0.533981
3        240000.0         15            0.145631  0.679612
4        260000.0          7            0.067961  0.747573
5        200000.0          6            0.058252  0.805825
6        360000.0          6            0.058252  0.864078
7        265000.0          6            0.058252  0.922330
8        220000.0          5            0.048544  0.970874
9        275000.0          5            0.048544  1.019417
10       210000.0          4            0.038835  1.058252
11       270000.0          4            0.038835  1.097087
12       216000.0          3            0.029126  1.126214
13       350000.0          3            0.029126  1.155340
14       230000.0          2            0.019417  1.174757
15       218000.0          2            0.019417  1.194175
16       252000.0          2            0.019417  1.213592
17       236000.0          2            0.019417  1.233010
18       204000.0          2            0.019417  1.252427
19       280000.0          2            0.019417  1.271845
20       276000.0          2            0.019417  1.291262
21       290000.0          1            0.009709  1.300971
22       336000.0          1            0.009709  1.310680
23       264000.0          1            0.009709  1.320388
24       287000.0          1            0.009709  1.330097
25       380000.0          1            0.009709  1.339806
26       295000.0          1            0.009709  1.349515
27       340000.0          1            0.009709  1.359223
28       233000.0          1            0.009709  1.368932
29       225000.0          1            0.009709  1.378641
30       255000.0          1            0.009709  1.388350
31       268000.0          1            0.009709  1.398058
32       320000.0          1            0.009709  1.407767
33       231000.0          1            0.009709  1.417476
34       285000.0          1            0.009709  1.427184
35       278000.0          1            0.009709  1.436893

# Skewness and Kurtosis in Data

## Definitions

### Skewness
Skewness measures the **asymmetry** of a distribution:
- **Positive skew**: Right tail is longer (mean > median)
- **Negative skew**: Left tail is longer (mean < median)
- **Zero skew**: Perfectly symmetrical (mean = median)

### Kurtosis
Kurtosis measures the **tailedness** (peakedness and tail thickness) of a distribution:
- **High kurtosis**: Heavy tails and sharp peak (more outliers)
- **Low kurtosis**: Light tails and flatter peak (fewer outliers)
- **Normal distribution** has kurtosis ≈ 3 (often reported as "excess kurtosis" = 0)


In [61]:
dataset["ssc_p"].skew()

-0.13264949031456927

In [62]:
dataset["ssc_p"].kurtosis()

-0.6075102066986688

In [63]:
def Univariate(dataset,quan):
    descriptive=pd.DataFrame(index=["Mean","Median","Mode","Min","Max","25%","50%","75%","100%","IQR",
                               "1.5IQR","LesserRange","GreaterRange","Skewness","Kurtosis"],columns=quan)
    
    for columnName in quan:
        descriptive[columnName]["Mean"]=dataset[columnName].mean()
        descriptive[columnName]["Median"]=dataset[columnName].median()
        descriptive[columnName]["Mode"]=dataset[columnName].mode()[0]
        descriptive[columnName]["Min"]=dataset[columnName].min()
        descriptive[columnName]["Max"]=dataset[columnName].max()
        descriptive[columnName]["25%"]=dataset.describe()[columnName]["25%"]
        descriptive[columnName]["50%"]=dataset.describe()[columnName]["50%"]
        descriptive[columnName]["75%"]=dataset.describe()[columnName]["75%"]
        descriptive[columnName]["100%"]=dataset.describe()[columnName]["max"]
        descriptive[columnName]["IQR"]=descriptive[columnName]["75%"]-descriptive[columnName]["25%"]
        descriptive[columnName]["1.5IQR"]=1.5*descriptive[columnName]["IQR"]
        descriptive[columnName]["LesserRange"]=descriptive[columnName]["25%"]- descriptive[columnName]["1.5IQR"]
        descriptive[columnName]["GreaterRange"]=descriptive[columnName]["75%"]+ descriptive[columnName]["1.5IQR"]
        descriptive[columnName]["Skewness"]=dataset[columnName].skew()
        descriptive[columnName]["Kurtosis"]=dataset[columnName].kurtosis()
    return descriptive

Univariate(dataset,quan)

sl_no     ssc_p      hsc_p   degree_p   etest_p     mba_p  \
Mean            108   67.3034    66.3347    66.3586   72.1006   62.2782   
Median          108        67         65         66        71        62   
Mode              1        62         63         65        60      56.7   
Min               1     40.89      42.75         50        50     51.21   
Max             215      89.4      91.15       88.5        98     77.89   
25%            54.5      60.6       60.9         61        60    57.945   
50%             108        67         65         66        71        62   
75%           161.5      75.7         73         72      83.5    66.255   
100%            215      89.4      91.15       88.5        98     77.89   
IQR             107      15.1       12.1         11      23.5      8.31   
1.5IQR        160.5     22.65      18.15       16.5     35.25    12.465   
LesserRange    -106     37.95      42.75       44.5     24.75     45.48   
GreaterRange    322     98.35      91.15       88.5    118.75     78.72   
Skewness          0 -0.132649   0.162611   0.204164  0.282308  0.313576   
Kurtosis       -1.2  -0.60751  0.0869008 -0.0974897  -1.08858 -0.470723   

                salary  
Mean            277649  
Median          265000  
Mode            300000  
Min             200000  
Max             390000  
25%             240000  
50%             265000  
75%             300000  
100%            390000  
IQR              60000  
1.5IQR           90000  
LesserRange     150000  
GreaterRange    390000  
Skewness        0.8067  
Kurtosis     -0.239837

## Analysis of Table

### Skewness Values Interpretation

| Column    | Skewness | Interpretation |
|-----------|----------|----------------|
| sl_no     | 0        | Perfect symmetry |
| ssc_p     | -0.13    | Slight left skew |
| hsc_p     | 0.16     | Slight right skew |
| degree_p  | 0.20     | Moderate right skew |
| etest_p   | 0.28     | Moderate right skew |
| mba_p     | 0.31     | Moderate right skew |
| salary    | 0.81     | Strong right skew |

### Kurtosis Values Interpretation

| Column    | Kurtosis | Interpretation |
|-----------|----------|----------------|
| sl_no     | -1.2     | Very flat distribution |
| ssc_p     | -0.61    | Platykurtic (flatter than normal) |
| hsc_p     | 0.09     | Nearly normal peakedness |
| degree_p  | -0.10    | Slightly platykurtic |
| etest_p   | -1.09    | Very flat distribution |
| mba_p     | -0.47    | Platykurtic |
| salary    | -0.24    | Slightly platykurtic |

## Visual Comparison

### Skewness Types Diagram
```
Left Skew (Negative)      Symmetrical       Right Skew (Positive)
      ▲                       ▲                      ▲
      │                       │                      │
      │                       │                      │
      │                       │                      │
      │                       │                      │
      └───────┐               │               ┌──────┘
              │               │               │
              └───────┐       │       ┌───────┘
                      └───────┴───────┘
```

### Kurtosis Types Diagram
```
Leptokurtic (Kurtosis > 3)    Mesokurtic (Kurtosis ≈ 3)    Platykurtic (Kurtosis < 3)
          ▲                            ▲                            ▲
          │                            │                            │
          │             /\            │            _____           │
          │            /  \           │           /     \          │
          │           /    \          │          /       \         │
          └──────────/      \─────────┴─────────/         \────────┘
```

## Key Observations from Your Data

1. **Salary Distribution**:
   - Strong right skew (0.81) - most salaries cluster at lower values with few high outliers
   - Slightly platykurtic (-0.24) - flatter than normal distribution

2. **Academic Performance (ssc_p, hsc_p, degree_p)**:
   - Mostly symmetrical to slightly right-skewed
   - Generally platykurtic (flatter than normal)

3. **sl_no (serial number)**:
   - Perfectly symmetrical (skewness = 0)
   - Very flat distribution (kurtosis = -1.2)

4. **Entrance Test (etest_p)**:
   - Moderate right skew (0.28)
   - Very flat distribution (-1.09)

## Practical Implications

1. **Salary Analysis**:
   - The right skew suggests most employees earn less than the mean salary
   - Consider using median for "typical" salary representation

2. **Academic Performance**:
   - The slight right skew in degree scores suggests more students perform above average
   - Flat distributions indicate no extreme outliers in academic scores

3. **Statistical Testing**:
   - Non-normal distributions may require non-parametric tests
   - Consider transformations for strongly skewed variables like salary


### Understanding Variance and Standard Deviation

In [67]:
def Univariate(dataset,quan):
    descriptive=pd.DataFrame(index=["Mean","Median","Mode","Min","Max","25%","50%","75%","100%","IQR",
                               "1.5IQR","LesserRange","GreaterRange","Skewness","Kurtosis", "Var","Std"],columns=quan)
    for columnName in quan:
        descriptive[columnName]["Mean"]=dataset[columnName].mean()
        descriptive[columnName]["Median"]=dataset[columnName].median()
        descriptive[columnName]["Mode"]=dataset[columnName].mode()[0]
        descriptive[columnName]["Min"]=dataset[columnName].min()
        descriptive[columnName]["Max"]=dataset[columnName].max()
        descriptive[columnName]["25%"]=dataset.describe()[columnName]["25%"]
        descriptive[columnName]["50%"]=dataset.describe()[columnName]["50%"]
        descriptive[columnName]["75%"]=dataset.describe()[columnName]["75%"]
        descriptive[columnName]["100%"]=dataset.describe()[columnName]["max"]
        descriptive[columnName]["IQR"]=descriptive[columnName]["75%"]-descriptive[columnName]["25%"]
        descriptive[columnName]["1.5IQR"]=1.5*descriptive[columnName]["IQR"]
        descriptive[columnName]["LesserRange"]=descriptive[columnName]["25%"]- descriptive[columnName]["1.5IQR"]
        descriptive[columnName]["GreaterRange"]=descriptive[columnName]["75%"]+ descriptive[columnName]["1.5IQR"]
        descriptive[columnName]["Skewness"]=dataset[columnName].skew()
        descriptive[columnName]["Kurtosis"]=dataset[columnName].kurtosis()
        descriptive[columnName]["Var"]=dataset[columnName].var()
        descriptive[columnName]["Std"]=dataset[columnName].std()
    return descriptive

In [68]:
Univariate(dataset,quan)

sl_no     ssc_p      hsc_p   degree_p   etest_p     mba_p  \
Mean              108   67.3034    66.3347    66.3586   72.1006   62.2782   
Median            108        67         65         66        71        62   
Mode                1        62         63         65        60      56.7   
Min                 1     40.89      42.75         50        50     51.21   
Max               215      89.4      91.15       88.5        98     77.89   
25%              54.5      60.6       60.9         61        60    57.945   
50%               108        67         65         66        71        62   
75%             161.5      75.7         73         72      83.5    66.255   
100%              215      89.4      91.15       88.5        98     77.89   
IQR               107      15.1       12.1         11      23.5      8.31   
1.5IQR          160.5     22.65      18.15       16.5     35.25    12.465   
LesserRange      -106     37.95      42.75       44.5     24.75     45.48   
GreaterRange      322     98.35      91.15       88.5    118.75     78.72   
Skewness            0 -0.132649   0.162611   0.204164  0.282308  0.313576   
Kurtosis         -1.2  -0.60751  0.0869008 -0.0974897  -1.08858 -0.470723   
Var              3870   117.228    112.064    53.6047   176.251   34.0284   
Std           62.2093   10.8272     10.586    7.32152    13.276   5.83338   

                  salary  
Mean              277649  
Median            265000  
Mode              300000  
Min               200000  
Max               390000  
25%               240000  
50%               265000  
75%               300000  
100%              390000  
IQR                60000  
1.5IQR             90000  
LesserRange       150000  
GreaterRange      390000  
Skewness          0.8067  
Kurtosis       -0.239837  
Var           2.9446e+09  
Std              54264.1


### Definitions

### Variance (σ²)
**Measures**: Spread of data points around the mean  

### Standard Deviation (σ)

**Measures**: Average distance from the mean (in original units)  

### Key Differences

| Characteristic | Variance | Standard Deviation |
|---------------|----------|---------------------|
| Units | Squared units (e.g., $²) | Original units (e.g., $) |
| Interpretation | Harder to interpret directly | More intuitive |
| Sensitivity | More sensitive to outliers | Less sensitive |

### When to Use Each

- **Use Variance** when:
  - Performing statistical tests (ANOVA, etc.)
  - Working with squared units in calculations

- **Use Standard Deviation** when:
  - Communicating results to non-technical audiences
  - Understanding real-world spread in original units

### Visualization

```python
import matplotlib.pyplot as plt

plt.figure(figsize=(10,5))
plt.hist(df['salary'], bins=30, alpha=0.7)
plt.axvline(mean, color='red', label='Mean')
plt.axvline(mean + std_dev, color='orange', linestyle='--', label='±1 STD')
plt.axvline(mean - std_dev, color='orange', linestyle='--')
plt.title("Salary Distribution with Standard Deviation")
plt.xlabel("Salary ($)")
plt.ylabel("Frequency")
plt.legend()
plt.show()
```


### **Normal Distribution Breakdown (Emprical Rule)** or Called Bell Shaped, Normal, Gaussian Distributions.

Your diagram shows a classic **bell curve** with standard deviation (σ) intervals from the mean (μ). Here's the complete explanation:

---

#### **1. Central Zone (μ ± 1σ)**
- **Range**: μ - 1σ to μ + 1σ  
- **Probability**: **68.2%** of data  
  (Shown as 34.1% + 34.1% in your diagram)  
- **Example**: If μ = 100 and σ = 15:  
  - 68% of values fall between **85-115**

#### **2. Intermediate Zone (μ ± 2σ)**
- **Range**: μ - 2σ to μ + 2σ  
- **Probability**: **95.4%** of data  
  (68.2% + 13.6% + 13.6%)  
- **Example**:  
  - 95% of values fall between **70-130**

#### **3. Outer Zone (μ ± 3σ)**
- **Range**: μ - 3σ to μ + 3σ  
- **Probability**: **99.7%** of data  
  (95.4% + 2.1% + 2.1% + 0.1% + 0.1%)  
- **Example**:  
  - 99.7% of values fall between **55-145**

---

### **Key Areas in Your Diagram**
| Label | Interpretation | Probability | Zone |
|-------|-----------------|-------------|------|
| μ | Mean (center) | N/A | Center |
| -30, -20, -10 | σ intervals (e.g., -3σ, -2σ, -1σ) | N/A | Left Tail |
| 0.1%, 2.1%, 13.6% | Right-tail probabilities | Cumulative beyond 1σ/2σ/3σ | Right Tail |
| 34.1% | Half of 68% zone | 34.1% per side | μ±1σ |

---

### **Practical Implications**
1. **Quality Control**:  
   - If a process output is normally distributed with μ = 100, σ = 5:  
     - **Red Flag**: Any value <85 (μ - 3σ) or >115 (μ + 3σ) is extremely rare (<0.3% chance).

2. **Grading Systems**:  
   - In exams with μ = 75, σ = 10:  
     - A grade (top 2.1%) ≈ score >95 (μ + 2σ)

3. **Finance**:  
   - Stock returns often follow this pattern:  
     - 95% of daily returns within ±2σ of average

---

### **Mathematical Confirmation**
For a true normal distribution:
```math
P(μ - σ ≤ X ≤ μ + σ) ≈ 68.27\% \\
P(μ - 2σ ≤ X ≤ μ + 2σ) ≈ 95.45\% \\
P(μ - 3σ ≤ X ≤ μ + 3σ) ≈ 99.73\%
```

---

### **Python Visualization**
```python
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import norm

x = np.linspace(-3, 3, 1000)
y = norm.pdf(x, 0, 1)  # μ=0, σ=1

plt.figure(figsize=(10,5))
plt.plot(x, y, 'k-')
plt.fill_between(x, y, where=(x >= -1) & (x <= 1), color='blue', alpha=0.3, label='68% (μ±1σ)')
plt.fill_between(x, y, where=(x >= -2) & (x <= 2), color='green', alpha=0.2, label='95% (μ±2σ)')
plt.fill_between(x, y, where=(x >= -3) & (x <= 3), color='red', alpha=0.1, label='99.7% (μ±3σ)')
plt.legend()
plt.title("Standard Normal Distribution")
plt.xlabel("Standard Deviations (σ)")
plt.ylabel("Probability Density")
plt.show()
```
---

### **Common Misconceptions**
1. **All bell curves are normal**: Only true if they follow the 68-95-99.7 rule exactly.  
2. **Outliers beyond 3σ don't exist**: They're rare but possible (0.3% probability).  
3. **Data must be normal**: Many real-world datasets are only approximately normal.  
